docs/enhancements/bmac-reboot-hosts-on-iso-change.md

```diff
@@ -0,0 +1,131 @@
+because the URLs won't match. A side benefit of ensuring that the ISO URL
+uniquely identifies the resource's contents, is that pull-through caches will
+have an easy time understanding when an file is no longer relevant.
+
+When assisted-service creates an ISO, it will capture and include in the ISO:
+* the name, namespace, UID, and generation of the InfraEnv
+* the name, namespace, UID, and generation of each NMStateConfig
+
+At runtime, those resource identifiers for the InfraEnv and the NMStateConfig
+it used will be passed by the agent to assisted-service and compared against
+the current resource generations to determine if they are obsolete.
+
+### assisted-service Agent controller
+
+When the Agent resource gets created, its status should include the
+above-described reference details for:
+* The InfraEnv
+* The NMStateConfig it utilized during its boot
+
+If those details indicate that any of the referenced resources have changed,
+then the controller will add a label to the Agent resource indicating such. It
+will also add a reference to each resource to an array in the Agent's Status.
+
+### Bare Metal Agent Controller
+
+The Bare Metal Agent Controller will watch for Agents to have that label and
+reboot the corresponding BareMetalHost with the latest ISO URL.
+
+The BMAC will also watch for the InfraEnv URL to change, and if there are any
+BareMetalHosts that were booted with an obsolete URL but do not yet have an
+Agent resource, those BareMetalHosts will be rebooted with the latest ISO URL.
+
+### User Stories
+
+#### Story 1
+
+As a user of Zero Touch Provisioning, when I add and modify NMStateConfig
+resources, unaffected hosts will not reboot. Only those hosts directly impacted
+will be rebooted with a current ISO URL.
+
+#### Story 2
+
+As a boot-it-yourself user, I will manually or with automation watch for Agents
+to have the obsolete artifact label and restart them with a fresh ISO.
+
+### Implementation Details/Notes/Constraints [optional]
+
+* Any Agent that has already started provisioning will be exempt from rebooting.
+
+### Risks and Mitigations
+
+
+## Design Details [optional]
+
+
+### Open Questions
+
+* Should the BMAC delete an Agent after it reboots the corresponding host?
+* What should the label's key and value be?
+* How exactly would the resource references get embedded into the ISO, and how
+  would the Agent communicate those back to assisted-service?
+
+### UI Impact
+
+
+### Test Plan
+
+
+## Drawbacks
+
+
+## Alternatives
+
+Each NMStateConfig resource could have a unique ISO URL. The InfraEnv would
+provide a base URL, and the NMStateConfig could be identified by query
+parameter or similar. This way, each host would get a unique URL and a unique
+ISO. When a particular NMStateConfig gets added or modified, it would
+implicitly only affect any BMH with that NMStateConfig's specific URL.
+
+Then BMAC would have to determine the correct NMStateConfig for a BareMetalHost
```

Comment:
BMAC should not care about the correct NMStateConfig. It should create a URL that embeds the ID of the host (BMC MAC?) .
The Assisted-Service, while streaming this URL should generate the appropriate ISO for that host ID